[Felissa Faustine] iP

felissaf opened the pull request:
# MyJournal 
MyJournal __*stores your busy schedule*__ and __*keeps all of your important deadlines and upcoming events*__!:laughing:

It is:
* Text-based
* Easy to use

What MyJournal can do:
- [x] Show your tasks (todo/event/deadline0
- [X] Mark completed tasks
- [X] Find tasks
- [X] Delete tasks

All you need to do is,
1. Download [here](https://github.com/felissaf/ip/releases)!
1. Add your tasks
1. You're all set!

As Benjamin Franklin said:
> For every minute spent in organising,
> an hour is earned.

This is the `main` method:
```java
public static void main(String[] args) {
    try {
        new MyJournal("./tasks.txt").run();
    } catch (IOException e) {
        e.printStackTrace();
    }
}
```

felissafaustine committed:
Added greet, echo and exit functions and changed the name of the chatbot

felissafaustine committed:
added add and list function, changed the introduction message

felissafaustine committed:
added Task class

felissafaustine committed:
added mark as done function

felissafaustine committed:
fixed some issues and changed the way it scans the input

felissafaustine committed:
Created deadline class

felissafaustine committed:
Created event class

felissafaustine committed:
Created todo class

felissafaustine committed:
Updated Task file

felissafaustine committed:
Added todo, event, deadline feature

felissafaustine committed:
Fixed todo, event, deadline inheritance and changed array to arraylist

felissafaustine committed:
Added EmptyDescriptionException and InvalidTypeException

felissafaustine committed:
Added InvalidTaskNumberException

felissafaustine committed:
Added delete feature

felissafaustine committed:
Add javadocs

felissafaustine committed:
Add testing using the I/O redirection technique

felissafaustine committed:
Change some formatting for the code and javadocs and edit the input file for more testing

felissafaustine committed:
Create taskAddPrint method

felissafaustine committed:
Add save and load file feature

felissafaustine committed:
Add date and time feature

felissafaustine committed:
Merge branches

# Conflicts:
#	src/main/java/MyJournal.java

felissafaustine committed:
Add package, Ui, Storage, Parser, TaskList

felissafaustine committed:
Add packages

felissafaustine committed:
Add JUnit Tests

felissafaustine committed:
Add feature to find task with keyword

felissafaustine committed:
Add javadocs

felissafaustine committed:
Follow coding standard

felissafaustine committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add javadocs

# Conflicts:
#	src/main/java/myjournal/Parser.java
#	src/main/java/myjournal/TaskList.java

felissafaustine committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add feature to find task with keyword

# Conflicts:
#	src/main/java/myjournal/Parser.java

felissafaustine committed:
Update welcome message

felissafaustine committed:
Change indentation in Parser class

felissafaustine committed:
Add gradle and do checkstyle

felissafaustine committed:
Fix gradle

felissafaustine committed:
Add GUI

felissafaustine committed:
Fix GUI

Xiaoyunnn commented on src/main/java/myjournal/Storage.java:
```diff
@@ -0,0 +1,92 @@
+        Scanner sc = new Scanner(file);
+        String currLine;
+        ArrayList<Task> tasks = new ArrayList<>();
+        while (sc.hasNextLine()) {
+            currLine = sc.nextLine();
+            String taskDescription = currLine.substring(5);
+            boolean isDone = currLine.charAt(2) == '1';
+            if (currLine.charAt(0) == 'T') {
+                Todo todo = new Todo(taskDescription);
+                todo.setState(isDone);
+                tasks.add(todo);
+            } else {
+                String[] arr = taskDescription.split(" ");
+                String description = "";
+                String time = "";
+                boolean check = false;
+                for (String s : arr) {
+                    if (s.equals("|")) {
+                        check = true;
+                    } else {
+                        if (check) {
+                            time = time + " " + s;
+                        } else {
+                            description = description + s + " ";
+                        }
+                    }
+                }
+                if (currLine.charAt(0) == 'E') {
+                    Event event = new Event(description, time);
+                    event.setState(isDone);
+                    tasks.add(event);
+                } else {
+                    Deadline deadline = new Deadline(description, time);
+                    deadline.setState(isDone);
+                    tasks.add(deadline);
+                }
+            }
+        }
+        return tasks;
+    }
```
This `load()` method might be too long. This [coding quality website](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html#avoid-long-methods) stated the following:

> Be wary when a method is longer than the computer screen, and take corrective action when it goes beyond 30 LOC (lines of code). The bigger the haystack, the harder it is to find a needle.

It also suggests that you should not have more than 3 levels of indentations. Maybe you can move the content in the `else` block into a separate method to make the code look cleaner?

felissaf replied:
Thank you for the suggestion!! I do feel like it’s too long but wasn’t really sure how to shorten it. I’ll try taking out the else block!:)

Xiaoyunnn commented on src/main/java/myjournal/Ui.java:
```diff
@@ -0,0 +1,80 @@
+public class Ui {
+    /**
+     * Prints the welcome message.
+     * @return
```
I think you forgot to add the description for `@return` here XD

felissaf replied:
Thanks for pointing it out!! I forgot to change the javadocs after changing my code oops XD

Xiaoyunnn commented on src/main/java/myjournal/Ui.java:
```diff
@@ -0,0 +1,80 @@
+                + "6. Type 'bye' to exit";
+    }
+
+    /**
+     * Prints out the statement after a task is added.
+     *
+     * @param taskList The list of all tasks.
+     */
+    public String taskAddPrint(TaskList taskList) {
+        return "Okay!! I've added the following task:\n"
+                + taskList.getTask(taskList.getSize() - 1) + "\n"
+                + "Now you have " + taskList.getSize() + " in the list";
+    }
```
I realised you seem to forget to add `@return` for most of your JavaDoc for methods that return something in this Ui class?

Xiaoyunnn commented on src/main/java/myjournal/Parser.java:
```diff
@@ -0,0 +1,292 @@
+    public static TaskList parseFind(Scanner line, TaskList tasks) {
+        if (!line.hasNext()) {
+            throw new EmptyDescriptionException("OOPS!!! Please tell me which word you want to find!!");
+        }
+        String find = line.next();
+        if (line.hasNext()) {
+            throw new InvalidWordException("Please enter only one word!!");
+        }
+        TaskList newList = new TaskList();
+        for (int i = 0; i < tasks.getSize(); i++) {
+            Task currTask = tasks.getTask(i);
+            Scanner taskName = new Scanner(currTask.getTaskName());
+            while (taskName.hasNext()) {
+                if (taskName.next().equals(find)) {
+                    newList.addTask(currTask);
+                }
+            }
+        }
+        return newList;
+    }
+
+    /**
+     * Parses user's input for the command "done".
+     *
+     * @param line The current line that is being parsed.
+     * @param tasks The list of tasks.
+     * @return A task which has been marked as done.
+     */
+    public static Task parseDone(Scanner line, TaskList tasks) {
+        if (!line.hasNextInt()) {
+            throw new InvalidTaskNumberException("OOPS!!! Please specify the task "
+                    + "that needs to be marked as done!");
+        }
+        int index = line.nextInt() - 1;
+        if (index >= tasks.getSize() || index < 0 || tasks.getTask(index) == null) {
+            throw new InvalidTaskNumberException("OOPS!!! Please enter a valid task number!");
+        }
+        tasks.getTask(index).setState(true);
+        return tasks.getTask(index);
+    }
```
Since the Exception thrown is not caught in the method body, shouldn't the method signature include `throws InvalidTaskNumberException` or `throws EmptyDescriptionException, InvalidWordException`

MGRL2201 replied:
Perhaps this could apply to parseDelete too?

Xiaoyunnn commented on src/main/java/myjournal/Parser.java:
```diff
@@ -0,0 +1,292 @@
+        String parsed = "";
+        while (line.hasNext()) {
+            String currWord = line.next();
+            if (isDate(currWord)) {
+                LocalDate date = LocalDate.parse(currWord);
+                String month = date.getMonth().getDisplayName(TextStyle.SHORT, Locale.ENGLISH);
+                String dateFormatted = date.getDayOfWeek().toString() + ", " + date.getDayOfMonth() + " " + month
+                        + " " + date.getYear();
+                parsed = parsed + " " + dateFormatted;
+            } else if (isTime(currWord)) {
+                LocalTime time = LocalTime.parse(currWord);
+                String beforeOrAfterNoon = time.getHour() >= 12 ? "pm" : "am";
+                int hour = time.getHour() == 0
+                        ? 12
+                        : time.getHour() >= 12
+                        ? time.getHour() - 12
+                        : time.getHour();
+                int min = time.getMinute();
+                String timeFormatted = (String.valueOf(hour).length() == 1 ? "0" + hour : hour)
+                        + ":" + (min < 10 ? "0" + min : min) + beforeOrAfterNoon;
+                parsed = parsed + " " + timeFormatted;
```
Perhaps you can consider using [DateTimeFormatter](https://www.baeldung.com/java-datetimeformatter) here to simplify your code?  😊

felissaf replied:
Okay I’ll try!! Thanks for the suggestion!:)

felissafaustine committed:
Fix welcome message and gradle

felissafaustine committed:
Change javafx version

felissafaustine committed:
Add exit after bye feature

MGRL2201 commented on src/main/java/myjournal/Storage.java:
```diff
@@ -0,0 +1,92 @@
+                Todo todo = new Todo(taskDescription);
+                todo.setState(isDone);
+                tasks.add(todo);
+            } else {
```
Perhaps to shorten your code, you could try using `.split(" \\| ")` which will split the line at the character |?

felissaf replied:
Thanks for the suggestions!!

MGRL2201 reviewed (commented):
Your code is well written and clean!

Agentum07 reviewed (commented):
Overall, the code looks great, good job!

I just have one suggestion - maybe create a "dukeException" class and have your other exception classes extend from that.

felissafaustine committed:
Add more javadocs

felissafaustine committed:
Improve code quality

Making the code less nested improves the code quality as code that is deeply nested is not very optimal.

The code is also shortened (but still has the same functionality) to improve readability and also improve code quality.
This is done by using DateTimeFormatter to shorten and make the getTimeDate method more efficient.

Some variable names are also changed to make it easier to understand the purpose of the variables.

felissafaustine committed:
Add assertions

Assertions are addded to Ui and Parser classes to stop the code from continue running after it encounters an error.

felissaf committed:
Merge pull request #1 from felissaf/branch-A-CodeQuality

Improve code quality

felissaf committed:
Merge pull request #2 from felissaf/branch-A-Assertions

Add assertions

felissafaustine committed:
Merge A-Assertions

felissafaustine committed:
Add Edit feature

When user want to change certain information about a task, they have to delete and re-add the task with the updated information, which is quite inconvenient.

This edit feature enables user to edit the time or description of any task.

This enables user to quickly change the description or the time of their task without having to delete and re-add the task.

felissafaustine committed:
Add image of Ui

felissafaustine committed:
Update image of Ui

felissaf committed:
Set theme jekyll-theme-hacker

felissafaustine committed:
Add user guide

felissafaustine committed:
Merge commit '94c1dea1aa14a2daa8e20896324e7476a81dbaf4'

* commit '94c1dea1aa14a2daa8e20896324e7476a81dbaf4':
  Set theme jekyll-theme-hacker

felissaf committed:
Set theme jekyll-theme-slate

felissafaustine committed:
Improve GUI

felissafaustine committed:
Fix GUI

felissafaustine committed:
Fix storage load issue

felissafaustine committed:
Change jar link

felissafaustine committed:
Change welcome message

felissafaustine committed:
Remove unnecessary test classes

felissafaustine committed:
Change event and deadline requirements

felissafaustine committed:
Remove printStackTrace

felissafaustine committed:
Delete unneccessary methods in MyJournal class

felissafaustine committed:
Combine exceptions

felissafaustine committed:
Change getTimeDate() method name

felissafaustine committed:
Edit import for Parser class

felissafaustine committed:
Remove exception

felissafaustine committed:
Delete unused import

felissafaustine committed:
Edit javadocs
